components/wmtk_components/embedding/internal/Embedding.cpp

```diff
@@ -0,0 +1,152 @@
+    igl::triangle::triangulate(
+        temp_vertices,
+        temp_edges,
+        outside_vertex,
+        "a0.1q",
+        m_vertices,
+        m_faces);
+
+    Eigen::MatrixXd V(m_vertices.rows(), 3);
+    for (int i = 0; i < m_vertices.rows(); ++i) {
+        V(i, 0) = m_vertices(i, 0);
+        V(i, 1) = m_vertices(i, 1);
+        V(i, 2) = 0;
+    }
+    // const std::filesystem::path data_dir =
+    //     "/home/zhouyuan/workplace/embedding_part/wildmeshing-toolkit/data";
+    // igl::writeOBJ(data_dir / "edge_test_mid.obj", V, m_faces);
+
+    spdlog::info("Triangulate Done!");
+
+    // need to connect the topology
+    // it would be easy, just check each edge and then move on
+    std::vector<std::vector<long>> adj_list;
+    igl::adjacency_list(m_faces, adj_list);
+
+    std::vector<std::pair<long, long>> temp_marked_edges;
+    std::vector<long> temp_marked_vertices;
+    for (const auto& e : m_marked_edges) {
+        // find marked E index before triangulate operation
+        long vid0 = e.first, vid1 = e.second;
+        Eigen::Vector2<double> p0, p1, dir;
+        p0 << m_vertices(vid0, 0), m_vertices(vid0, 1);
+        p1 << m_vertices(vid1, 0), m_vertices(vid1, 1);
+        long cur_vid = vid0;
+        temp_marked_vertices.push_back(cur_vid);
+        // our start vertex is p0, and constantly head to p1, and stop when encounter p1
+        while (cur_vid != vid1) {
+            // spdlog::info("{}->{}", cur_vid, vid1);
+            Eigen::Vector2<double> cur_p;
+            cur_p << m_vertices(cur_vid, 0), m_vertices(cur_vid, 1);
+            dir = (p1 - cur_p).normalized();
+            const std::vector<long>& neighbour_v_list = adj_list[cur_vid];
+            double value_dir = -1;
+            long next_vid = -1;
+            // for each step, just foward the direction with less value of dot operation(sin)
+            for (const long neighbour_id : neighbour_v_list) {
+                Eigen::Vector2<double> p, cur_dir;
+                p << m_vertices(neighbour_id, 0), m_vertices(neighbour_id, 1);
+                cur_dir = (p - cur_p).normalized();
+                if (value_dir < cur_dir.dot(dir)) {
+                    value_dir = cur_dir.dot(dir);
+                    next_vid = neighbour_id;
+                }
+            }
+            long vid0_ = cur_vid, vid1_ = next_vid;
+            if (vid0_ > vid1_) std::swap(vid0_, vid1_);
+            temp_marked_edges.push_back(std::pair<long, long>(vid0_, vid1_));
+            temp_marked_vertices.push_back(next_vid);
+            cur_vid = next_vid;
+        }
+    }
+    std::sort(temp_marked_vertices.begin(), temp_marked_vertices.end());
+    temp_marked_vertices.erase(
+        std::unique(temp_marked_vertices.begin(), temp_marked_vertices.end()),
+        temp_marked_vertices.end());
+    m_marked_edges = temp_marked_edges;
+    m_marked_vertices = temp_marked_vertices;
+
+    m_vertex_tags = std::vector<long>(m_vertices.rows(), options.embedding_tag_value);

```

Comment:
I think these tags should be edge attributes, am I right?

Reply:
m_marked_edge only contains the index of the edges on the input edges.
Since the index is not match to the toolkit's index, I convert these attribute into the trimesh in the file embedding.cpp